File: server/datasets/java_source-code/Multiple-C2/organic_5.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #5
 * Object-oriented solver: a KnapsackSolver class wraps items and capacity.
 */

class KnapsackSolver {
    int[] weights;
    int[] values;
    int capacity;
    int n;

    public KnapsackSolver(int[] weights, int[] values, int capacity) {
        this.weights = weights;
        this.values = values;
        this.capacity = capacity;
        this.n = weights.length;
    }

    public int solve() {
        int[][] dp = new int[this.n + 1][this.capacity + 1];

        for (int i = 1; i <= this.n; i++) {
            int weightI = this.weights[i - 1];
            int valueI = this.values[i - 1];
            for (int cap = 0; cap <= this.capacity; cap++) {
                if (weightI > cap) {
                    dp[i][cap] = dp[i - 1][cap];
                } else {
                    dp[i][cap] = Math.max(dp[i - 1][cap], dp[i - 1][cap - weightI] + valueI);
                }
            }
        }
        return dp[this.n][this.capacity];
    }

    public static void main(String[] args) {
        KnapsackSolver solver = new KnapsackSolver(new int[]{4, 5, 1, 3}, new int[]{10, 40, 10, 30}, 10);
        System.out.println(solver.solve());
    }
}

File: server/datasets/controlled_benchmarks/python/multiple_files/multiple_c2/organic_5.py
"""
0/1 Knapsack (DP) - Organic Submission #5
Object-oriented solver: a KnapsackSolver class wraps the items and capacity.
"""


class KnapsackSolver:
    def __init__(self, weights, values, capacity):
        self.weights = weights
        self.values = values
        self.capacity = capacity
        self.n = len(weights)

    def solve(self):
        dp = [[0] * (self.capacity + 1) for _ in range(self.n + 1)]

        for i in range(1, self.n + 1):
            weight_i = self.weights[i - 1]
            value_i = self.values[i - 1]
            for cap in range(self.capacity + 1):
                if weight_i > cap:
                    dp[i][cap] = dp[i - 1][cap]
                else:
                    dp[i][cap] = max(dp[i - 1][cap], dp[i - 1][cap - weight_i] + value_i)

        return dp[self.n][self.capacity]


if __name__ == "__main__":
    solver = KnapsackSolver(weights=[4, 5, 1, 3], values=[10, 40, 10, 30], capacity=10)
    print(solver.solve())
